### PR TITLE
by pass missing value of NaN

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
If the missing values is Nan, all the calculation will be Nan. It is not necessary to swap or test
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by weiyuan-jiang** [comment]
This PR is addressing the issue https://github.com/ufs-community/ufs-weather-model/issues/1259

**Comment by mathomp4** [comment]
I can say this is zero-diff for a C24 run of GEOSgcm. I'll let @tclune or someone more knowledgable than I comment on the code change.

**Comment by weiyuan-jiang** [comment]
> I don't see how this change has anything to do with NaN. The change merely verifies that if a fill is specified then it is of a numeric type. (I.e., it will ignore fills that are strings or bools.) What am I missing?

This line crashes without the fix. 
https://github.com/GEOS-ESM/MAPL/blob/d46c2ed476db26b4d21ba52da5cb4256b1a6cf0e/griddedio/GriddedIO.F90#L1092

I suspect this is a bug from ESMF.  Just two lines above,  I test the return type of the '_FillValue',  it is ESMF_TYPEKIND_R4 , but  ESMF crashes when trying to get it.

**Comment by weiyuan-jiang** [comment]
>

**Comment by tclune** [comment]
I don't see how your change would affect that case.   R4 matches so `has_missing_value` should be `.true.` before and after your change.   If the code change affects results, it seems to suggest memory corruption.

**Comment by weiyuan-jiang** [comment]
> @bena-nasa , I think it is fine to bypass the Nan values even there is Nan. Any calculation relults that Nan is involved would be Nan too.

Our Metadata does not take "Nan" as a numerical type. So there is no missing data if it is Nan. I think it is fine because the results would be Nan if Nan is involved in the calculation

**Comment by weiyuan-jiang** [comment]
I have another theory. When ESMF checks the type of attribute '_FillValue', it actually checks the type of the variable. They are supposed to be the same type anyway. Whey trying to get it,  it crashes because it is a string ( Nan or Nanf)

**Comment by tclune** [comment]
Your theory is plausible, but I still don't see how the changes here have any effect.

**Comment by weiyuan-jiang** [comment]
pfio can set and get Nan attribute,  but ESMF does not seem to work.  But I think it is safe to by pass Nan.

File: griddedio/GriddedIO.F90 [hunk 1]
```diff
@@ -1013,6 +1013,10 @@ subroutine request_data_from_file(this,filename,timeindex,rc)
              this%read_collection_id, fileName, trim(names(i)), &
              & ref, start=localStart, global_start=globalStart, global_count=globalCount)
         deallocate(localStart,globalStart,globalCount)
+        ! if is Nan, do not set '_FillValue'.
+        ! The pair ESMF_AttributeSet and ESMF_AttributeGet cannot handle Nan
+        if (missing_value /= missing_value) cycle
```
Comment (tclune) [inline griddedio/GriddedIO.F90:1018]: Ok.   

1. Perhaps we should return an error if it is found to be NaN in this manner?   Better to fail here then when doing a regrid with such values later on.  
2. Please open a ticket with esmf support to see if this is something that they can fix on their end.
3. This mechanism for detecting NaN is just obscure enough that it is probably worth it to introduce a helper function `is_NaN(x)`  and then call that instead.   I would not ask, except you have to modify the code still to handle the error condition.

**Review by tclune — CHANGES_REQUESTED** [review]
See inline comment.

File: griddedio/GriddedIO.F90 (hunk 1 above)
Comment (mathomp4) [reply griddedio/GriddedIO.F90:1018]: If it is a "real" nan, `ieee_is_nan()` could do it for us. We use it in various places in GEOS. (Though for historical purposes we often rename it as `isnan`, a la https://cs.github.com/GEOS-ESM/GEOSgcm_GridComp/blob/3e52659a4baa55aea5b0c937afcbfe90f0fafd92/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/mk_CatchCNRestarts.F90#L199)

**Comment by weiyuan-jiang** [comment]
If  regrid gets trouble when there is Nan, this PR won't work. Nan needs to be replaced somewhere.

File: griddedio/GriddedIO.F90 (hunk 1 above)
Comment (weiyuan-jiang) [reply griddedio/GriddedIO.F90]: The Nan now is changed to MAPL_UNDEF. But I think it is fine not to replace it.  The results will be Nan if a Nan is involved in calculation

Comment (tclune) [reply griddedio/GriddedIO.F90]: @weiyuan-jiang please elaborate on what you wrote above.    NaN is never MAPL_UNDEF, but I think you meant that the FillValue is now changed to MAPL_UNDEF.   This is not a change we are permitted to make though.

Comment (bena-nasa) [reply griddedio/GriddedIO.F90]: I just don't know what the right thing to do here is, we could fail and return an error if the missing value is nan. On the other hand if we do not, the latest commit from weiyuan seems like all we can do. We can't let the nan's propagate to a real calculation. At least if the Nan's are replaced with MAPL_UNDEF which we using internal no matter what the missing value is those points will be skipped without leading to an error. I would approve but I will defer to @tclune.

Comment (weiyuan-jiang) [reply griddedio/GriddedIO.F90]: @tclune I am not sure about "This is not a change we are permitted to make though".  I think we simply replace Nan in the memory with MAPL_UNDEF for the convenience of calculation

Comment (tclune) [reply griddedio/GriddedIO.F90]: ok - trying to look at the code to see where this has an impact.  The CYCLE seems to suggest we don't need to tell ESMF the fill value in the first place.

Comment (weiyuan-jiang) [reply griddedio/GriddedIO.F90]: How about saving the metadata for swap call when request_data_from_file. Is it safe?  @bena-nasa  , please see my new change

**Comment by weiyuan-jiang** [comment]
I have verified this PR resolves the Nan issue from ufs @bena-nasa @tclune

**Review by tclune — DISMISSED** [review]

**Comment by weiyuan-jiang** [comment]
@mathomp4 This CMakeLists.txt is working on Orion.  If it is working here, lets merge it.

**Review by tclune — APPROVED** [review]

**Comment by mathomp4** [comment]
Okay. Let me make an update to CMakeLists.txt with a comment.

**Review by mathomp4 — APPROVED** [review]
@tclune has approved. This is just a "re-approval" after my CMake comment.

**Comment by jkbk2004** [comment]
@mathomp4 are you going to release a new patched version of the mapl-2.21.3 out of this pr?

**Comment by mathomp4** [comment]
> @mathomp4 are you going to release a new patched version of the mapl-2.21.3 out of this pr?

@jkbk2004 I think once we get in #1580, develop should be in a good place. But let me ping the team to see if anything else needs in. If not, I'll work on getting something out today (probably MAPL 2.22).

**Pull request merged**